doc/modules/ROOT/pages/python-runtime/knowledge-graph-embeddings.adoc

```diff
@@ -0,0 +1,283 @@
+
+* Stochastic Local Closed World Assumption (sLCWA) assumes that all observed facts are true.
+Some corrupted triplets are false and some are true.
+The number of corrupted triplets for each true triplet is set by the `negative_sampling_size` parameter.
+
+Knowledge graph embedding models are trained under the sLCWA assumption in current implementation.
+
+
+=== Loss function
+The loss function is crucial for guiding the training process of KGE models.
+It determines how the difference between predicted and actual values is calculated and minimized.
+There are several loss functions that can be used for training KGE models, see below for more details.
+
+==== Margin ranking loss
+
+Margin ranking loss is a pairwise loss function that compares the scores based on the difference between the scores of a positive triple and a negative triple.
+When negative sampling size is more than one, the loss is computed for positive triple and each negative triple and the average loss is computed.
+
+image::python-runtime/mrl.svg[width=300]
+image::python-runtime/delta-value.svg[width=400]
+
+
+==== Negative Sampling Self-Adversarial Loss
+
+Negative Sampling Self-Adversarial Lossfootnote:[Sun, Zhiqing, et al. "Rotate: Knowledge graph embedding by relational rotation in complex space." arXiv preprint arXiv:1902.10197 (2019).] is a setwise loss function that compares the scores based on the difference between the scores of a positive triple and a set of negative triples.
+`loss_function_kwargs` can be used to set the `adversarial_temperature` and `margin` parameters.
+
+
+=== Optimizer
+
+Any pytorch optimizer can be used for training the model.
+To use non-default optimizer, specify the optimizer class name as a string in the `optimizer` parameter.
+All optimizer parameters except `params` can be passed as `optimizer_kwargs`.
+
+=== Negative sampling
+
+Loss function requires negative samples to compute the loss.
+The number of negative samples per positive sample is controlled by the `negative_sampling_size` parameter.
+When `use_node_type_aware_sampler` is set to `True`, negative nodes are sampled with the same label as the corresponding positive node.
+With or without node type awareness, the negative samples are sampled uniformly at random from the graph.
+
+=== Learning rate scheduler
+
+Any pytorch learning rate scheduler can be used for training the model.
+To use non-default learning rate scheduler, specify the scheduler class name as a string in the `lr_scheduler` parameter.
+All scheduler parameters except `optimizer` can be passed as `lr_scheduler_kwargs`.
+
+=== Inner normalisation
+
+In the original `TransE` paperfootnote:[Bordes, Antoine, et al. "Translating embeddings for modeling multi-relational data." Advances in neural information processing systems 26 (2013).]
+in `Algorithm 1`, line 5, the entity embeddings are normalized to have `Lp` norm of 1.
```

Comment:
What is the value of `p` in this case?

Reply:
Ok! Can the user set which `p` they want?

Reply:
Yes, it's the same as `p_norm` value, I'll add information about.

Reply:
ok 👍